### ISSUE TITLE
do expressions should allow return statements

### ISSUE THREAD

**Issue by jamiebuilds** [babel/babel#5233]
https://twitter.com/sebmarkbage/status/825197254252257280

```js
function method() {
  let a = do {
    if (true) return 1;
    2;
  };
  return a;
}

let result = method();

assert(result === 1);
```

**Comment by hzoo** [comment]
What would this transform into? I think we need to allow returns in do expressions in babylon since it's only in functions atm.

**Comment by thejameskyle** [comment]
This is going to be a more difficult transform for sure. 

It's gonna be something like this:

**Input:**

```js
function method() {
  let a = do {
    if (true) return 1;
    2;
  };
  return a;
}
```

**Output:**

```js
function method() {
  var _earlyReturn = false; // need for returning later on
  let a;
  var _result = function() {
    if (true) return (_earlyReturn = true, 1); // , expression is easiest way to accomplish this
    return 2;
  }();
  if (_earlyReturn) return _result;
  a = _result; // need this like in case _result.prop << might not exist, might be getter
  return a;
}
```

**Comment by bakkot** [comment]
In the proposal, `break` and `continue` crossing the do-expression boundary would also be legal, a la:

```js
for (let a of b) {
  c.push(do { if (a === null) break; a; });
}
```

(Personally I'm not so sure allowing non-exceptional control flow in do-expressions is a good idea, but it's what the proposal calls for.)

**Comment by loganfsmyth** [comment]
Allowing `return`, `continue`, `break`, and `throw` makes sense from my view. Do expressions should behave like any other block statement.

**Comment by thejameskyle** [comment]
do expressions are a potential precursor to making every statement in JS into an expression. `return`, `break`, `continue`, etc all need to work. @loganfsmyth is right, there's little difference from a block statement.

**Comment by bakkot** [comment]
I disagree, but this issue probably isn't the place for that conversation. I just wanted to note that it's what the proposal currently calls for.
